pkg/resource/deploy/plan_executor.go

```diff
@@ -0,0 +1,224 @@
+// Copyright 2016-2018, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package deploy
+
+import (
+	"context"
+	"sync/atomic"
+
+	"github.com/pkg/errors"
+	"github.com/pulumi/pulumi/pkg/util/cancel"
+	"github.com/pulumi/pulumi/pkg/util/logging"
+)
+
+const (
+	// Utility constant for easy debugging.
+	planExecutorLogLevel = 4
+)
+
+var (
```

Comment:
A huge struggle I have with Go in general is that it's basically impossible to tell the set of errors that can come out of a method call. I like pre-allocating errors so that I can document ahead-of-time exactly the sorts of errors that are allowed to come out of a particular package.

In this case, the stack trace capturing is not interesting, since these three errors are all coming from `PlanExecutor.Execute`. I'm most interested in being able to inspect these errors higher up the stack and avoid bad error messages like "an error occured while advancing the preview", which bubbles up to the top because we're using stringly-typed errors for control flow.